### ISSUE TITLE
Incorrect sampling from public data in cross-dataset InstaHide

### ISSUE THREAD

**Issue by carlini** [Hazelsuko07/InstaHide#2]
The paper states it is very important for "for Cross-dataset InstaHide [to] use a very large public dataset" to increase the security of the system.

The code that samples from this dataset, however, only uses the first `batch_size` samples. This means even if the dataset has 1M samples, if the batch size is 128, then the current code will only mix in the first 128 examples from the public data.

https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L143-L150

**Comment by Hazelsuko07** [comment]
Thanks for the comments. 

In our implementation, the 'batch_size' in the `mixup_data()` function is actually the same as the size of the private dataset (i.e., `inputs`), since we load all the training data once. 

https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L159-L166

Since the public dataset (i.e., `inputs_help`) is larger than the private one, doing random indexing inside the `mixup_data()` function assigns each mixing a different subset of public samples.

We implement our own batching schedule inside the `train()` function.
https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L174-L175

**Comment by carlini** [comment]
Unfortunately you're not correct.

You're right in the details everything you've said, but all of that is independent of the bug. The code only sample the first private_data_size samples.

This is a critical bug for the contest: because the batch size on the contest is 100, it implies that you are only mixing up 100 public images with the private contest images.

More worryingly, I now question if your cross-dataset scheme works. Because your defense has no proof of security, and it's only the implementation that shows it's secure, I wonder if the buggy code you've developed gave you a false sense of security.


Here's two ways to check this.

1. Reading the code

First let's figure out what the state of the world loks like at the start of line 383.

https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L383-L384

Here, we have
- `trainloader_help` which will generate the public dataset with batches of size `len(trainset_help)`
- `inputs_help` which, therefore, has shape `(len(trainset_help), *, *, *)`,
- `trainloader` which will generate training dataset batches of size `len(trainset)`

Concretely, if we were setting up the contest, this means we have inputs_help is ~1M samples and `trainloader` will generate 100 samples.

Now let's look at the function `generate_sample` on line 159

https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L159-L166

This is going to loop over the trainloader (which is fine because the private batch size is the private dataset size), and call `mixup_data` on input with shape
- `inputs` which will have data of shape `(len(trainset), *, *, *)`
- `target` which is irrelavent for the bug
- `inputs_help` which again still has shape `(len(trainset_help), *, *, *)`

So now we step into `mixup_data` on line 118

https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L118

We're binding `inputs` to `x`, and `inputs_help` to `x_help`. Now keep going to where we compute batch_size on line 143

https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L143

And we see that the variable `batch_size` is going to be equal to the size of the *private* dataset.

Therefore, on the next line, when we generate the random permutation

https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L144

this will be a permutation only of the in the range from `0` to `batch_size`, which is, again, the private batch size.

As a result of all of this, when we finally get to sampling from x_help, on line 150

https://github.com/Hazelsuko07/InstaHide/blob/8048b06a51d745982806a09fb46aca0b063dd3cb/train_cross.py#L150

We end up choosing only the first [0..batch_size-1] samples. This is not what the paper describes, and is incorrect.



2. Running the code

The above analysis was somewhat long. It's possible that maybe I've made a mistake. Let's just run the code and find out.

To do this, we make the following patch:


```diff
diff --git a/train_cross.py b/train_cross.py
index 4abb7b5..4d247a7 100755
--- a/train_cross.py
+++ b/train_cross.py
@@ -142,6 +142,7 @@ def mixup_data(x, y, x_help, use_cuda=True):
     for i in range(1, args.klam):
         batch_size = x.size()[0]
         index = torch.randperm(batch_size).to(device)
+        print("Perm", index.shape, x_help.shape, x_help[0][0][0])
         if i < inside_cnt:
             # mix private samples
             mixed_x += vec_mul_ten(lams[:, i], x[index, :])
@@ -161,6 +162,7 @@ def generate_sample(trainloader, inputs_help):
     for _, (inputs, targets) in enumerate(trainloader):
         if use_cuda:
             inputs, targets = inputs.cuda(), targets.cuda()
+        print("Feed", inputs.shape)
         mix_inputs, mix_targets, lams = mixup_data(
             inputs, targets.float(), inputs_help, use_cuda)
     return (mix_inputs, mix_targets, lams)
@@ -378,10 +380,13 @@ def main():
     for _, (inputs_help, targets) in enumerate(trainloader_help):
         if use_cuda:
             inputs_help = inputs_help.cuda()
+    inputs_help = torch.randn((100000,3,32,32))
+    inputs_help[50000:,:,:,:] = np.nan

     for epoch in range(start_epoch, args.epoch):
         mix_inputs_all, mix_targets_all, lams = generate_sample(
             trainloader, inputs_help)
+        print('nan count', np.sum(np.isnan(mix_inputs_all.numpy())))
         train_loss, _ = train(
             net, optimizer, mix_inputs_all, mix_targets_all, lams, epoch)
         test_loss, test_acc1, = test(
```

Walking through what I've done:
- First, every time we generate the permutation, I'm printing out the shape of the permutation and the size of the public dataset, along with the first entry in the public dataset.
- Then, every time we generate a sample, I pass the size of the input shape
- After that, I'm going to create a new `inputs_help` data that is just random values for the first 50k entries and NaN for the next 50k
- Finally, after we create the mixed data, I'm counting the number of NaNs.

So as described in the README, I'm going to run this

```bash
python3 train_cross.py --mode instahide --klam 6 --data cifar10 --pair --dom 0.3 --help_dir /tmp/img
```

And then I get the following output to establish a baseline.

```
==> Number of lambdas: 6
==> Preparing data..
Files already downloaded and verified
Files already downloaded and verified
==> Building model..
Feed torch.Size([50000, 3, 32, 32])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.8004,  1.3091, -0.1191,  0.7643, -0.5947,  0.6844, -0.6019, -1.5432,
        -0.1995,  0.9499, -0.1711,  0.1157, -1.1863,  1.7327, -1.9590, -0.3115,
        -0.3593,  1.4589, -0.1720, -0.7931, -0.4082, -0.6312, -1.7526, -0.1888,
        -0.4829,  0.6571, -0.4219, -1.8851,  1.9860,  0.7269, -2.0555,  1.2599])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.8004,  1.3091, -0.1191,  0.7643, -0.5947,  0.6844, -0.6019, -1.5432,
        -0.1995,  0.9499, -0.1711,  0.1157, -1.1863,  1.7327, -1.9590, -0.3115,
        -0.3593,  1.4589, -0.1720, -0.7931, -0.4082, -0.6312, -1.7526, -0.1888,
        -0.4829,  0.6571, -0.4219, -1.8851,  1.9860,  0.7269, -2.0555,  1.2599])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.8004,  1.3091, -0.1191,  0.7643, -0.5947,  0.6844, -0.6019, -1.5432,
        -0.1995,  0.9499, -0.1711,  0.1157, -1.1863,  1.7327, -1.9590, -0.3115,
        -0.3593,  1.4589, -0.1720, -0.7931, -0.4082, -0.6312, -1.7526, -0.1888,
        -0.4829,  0.6571, -0.4219, -1.8851,  1.9860,  0.7269, -2.0555,  1.2599])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.8004,  1.3091, -0.1191,  0.7643, -0.5947,  0.6844, -0.6019, -1.5432,
        -0.1995,  0.9499, -0.1711,  0.1157, -1.1863,  1.7327, -1.9590, -0.3115,
        -0.3593,  1.4589, -0.1720, -0.7931, -0.4082, -0.6312, -1.7526, -0.1888,
        -0.4829,  0.6571, -0.4219, -1.8851,  1.9860,  0.7269, -2.0555,  1.2599])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.8004,  1.3091, -0.1191,  0.7643, -0.5947,  0.6844, -0.6019, -1.5432,
        -0.1995,  0.9499, -0.1711,  0.1157, -1.1863,  1.7327, -1.9590, -0.3115,
        -0.3593,  1.4589, -0.1720, -0.7931, -0.4082, -0.6312, -1.7526, -0.1888,
        -0.4829,  0.6571, -0.4219, -1.8851,  1.9860,  0.7269, -2.0555,  1.2599])
nan count 0
```

And no we see that the bug exists:
- We are passing 50k samples as the batch, and 100k samples as the public size, but only using the first 50k of them.
- The nan count is 0 because we never sample from any sample after the first 50k public images.

Alright, now let's change just one line, and instead of setting samples [50000:] to NaN, let's set [40000:] to NaN. Just to really make sure that this is what we expect

```
==> Number of lambdas: 6
==> Preparing data..
Files already downloaded and verified
Files already downloaded and verified
==> Building model..
Feed torch.Size([50000, 3, 32, 32])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.0693,  0.0514,  2.0108,  1.4903,  0.6967,  1.9885, -1.2817, -0.3190,
         0.9387,  0.0340, -2.2145, -0.3056,  0.1339,  0.7536, -0.8395,  2.0170,
        -0.8305,  1.0033,  0.3805,  2.8081, -0.7264, -0.0300, -1.5406,  0.0092,
        -0.2725,  0.2183,  0.8901,  1.0008,  2.1819,  0.1949,  0.7425, -1.9489])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.0693,  0.0514,  2.0108,  1.4903,  0.6967,  1.9885, -1.2817, -0.3190,
         0.9387,  0.0340, -2.2145, -0.3056,  0.1339,  0.7536, -0.8395,  2.0170,
        -0.8305,  1.0033,  0.3805,  2.8081, -0.7264, -0.0300, -1.5406,  0.0092,
        -0.2725,  0.2183,  0.8901,  1.0008,  2.1819,  0.1949,  0.7425, -1.9489])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.0693,  0.0514,  2.0108,  1.4903,  0.6967,  1.9885, -1.2817, -0.3190,
         0.9387,  0.0340, -2.2145, -0.3056,  0.1339,  0.7536, -0.8395,  2.0170,
        -0.8305,  1.0033,  0.3805,  2.8081, -0.7264, -0.0300, -1.5406,  0.0092,
        -0.2725,  0.2183,  0.8901,  1.0008,  2.1819,  0.1949,  0.7425, -1.9489])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.0693,  0.0514,  2.0108,  1.4903,  0.6967,  1.9885, -1.2817, -0.3190,
         0.9387,  0.0340, -2.2145, -0.3056,  0.1339,  0.7536, -0.8395,  2.0170,
        -0.8305,  1.0033,  0.3805,  2.8081, -0.7264, -0.0300, -1.5406,  0.0092,
        -0.2725,  0.2183,  0.8901,  1.0008,  2.1819,  0.1949,  0.7425, -1.9489])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.0693,  0.0514,  2.0108,  1.4903,  0.6967,  1.9885, -1.2817, -0.3190,
         0.9387,  0.0340, -2.2145, -0.3056,  0.1339,  0.7536, -0.8395,  2.0170,
        -0.8305,  1.0033,  0.3805,  2.8081, -0.7264, -0.0300, -1.5406,  0.0092,
        -0.2725,  0.2183,  0.8901,  1.0008,  2.1819,  0.1949,  0.7425, -1.9489])
nan count 90777600
```

And yes, the nan count is huge.

So, if the code really was using the full public dataset size, then we should have seen some NaNs show up the first time around. They don't; therefore it's buggy.

There is one final thing we could hope for. Maybe somewhere the `inputs_help` dataset gets shuffled, and each time we generate the data we only use the first 50k, but the first 50k changes each time. 

This also does not happen: if you let the code run for two epochs of training, you'll see that the value of the first row of the public dataset doesn't change between epochs. (I ran it for 5 epochs to be really sure, and rule out that 1-in-a-billion case.)

```
==> Number of lambdas: 6
==> Preparing data..
Files already downloaded and verified
Files already downloaded and verified
==> Building model..
Feed torch.Size([50000, 3, 32, 32])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
nan count 0

Epoch: 0
 [======================================== 79/79 =====================================>.]  Step: 199ms | Tot: 19s13ms | Loss: 2.612 | Acc: 12.910% (1291/10000)
Saving..
Feed torch.Size([50000, 3, 32, 32])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
Perm torch.Size([50000]) torch.Size([100000, 3, 32, 32]) tensor([ 0.3131,  0.0414,  0.9006,  1.0899,  1.7320, -0.1320, -0.1343, -0.5916,
         1.1680, -0.8395, -0.5007, -0.4393,  0.2789, -1.4815, -1.2494,  0.6056,
        -0.5282, -1.1795,  0.4947,  1.3475, -1.3258, -1.2151,  0.7185, -0.6096,
        -0.6580,  1.8152, -0.1894,  1.2226, -0.2670, -1.2642, -0.0880, -0.4883])
nan count 0
```

Again, looks like no.

This is all that I can think of to show you that this is indeed a real bug. Please let me know if there's something I missed.

**Comment by Hazelsuko07** [comment]
Nicholas,

Thanks a lot for walking through this issue. Yes, you are right about this: the previous version only samples the first private_data_size images from the public dataset. 

Just committed a quick fix in adc1b45f24b6f5017020686e23e64560b904fa2c by permuting the public dataset (`inputs_help`) per epoch. Later, we will optimize the sampling process for better efficiency.

Here is the test accuracy (across 5 runs) on CIFAR-10 after the fix: 90.4±0.2 for k=4 (was 90.7±0.2), and 89.7 ± 0.6 for k=6 (was 89.8±0.6). We will update the cross-dataset results for other benchmarks in an updated arxiv version.

> This is a critical bug for the contest: because the batch size on the contest is 100, it implies that you are only mixing up 100 public images with the private contest images.

Yes. We are heading to update the contest dataset.
